pkg/query-service/app/querier/helper.go

```diff
@@ -284,18 +384,28 @@ func (q *querier) runBuilderExpression(
 	}
 	mergedSeries := mergeSerieses(cachedSeries, missedSeries)
 
+	var mergedSeriesData []byte
+	missedSeriesLen := len(missedSeries)
+	if missedSeriesLen > 0 && !params.NoCache && q.cache != nil {
+		// caching the data
+		var err error
+		mergedSeriesData, err = json.Marshal(mergedSeries)
+		if err != nil {
+			zap.S().Error("error marshalling merged series", zap.Error(err))
+			return
```

Comment:
You don't want to return here because there was some issue with marshalling the series. The return here is not the same as the return after `ch <- channelResult`. Just log the error and proceed. And handle the `mergedSeriesData` before `q.cache.Store`